config/backend-config/multitenant_workspaces.go

```diff
@@ -136,41 +139,46 @@ func (workspaceConfig *MultiTenantWorkspacesConfig) getFromAPI(ctx context.Conte
 		}
 		sourcesJSON.Sources = append(sourcesJSON.Sources, workspaceConfig.Sources...)
 	}
-	workspaceConfig.workspaceWriteKeysMapLock.Lock()
-	workspaceConfig.writeKeyToWorkspaceIDMap = writeKeyToWorkspaceIDMap
-	workspaceConfig.sourceToWorkspaceIDMap = sourceToWorkspaceIDMap
-	workspaceConfig.workspaceIDToLibrariesMap = workspaceIDToLibrariesMap
-	workspaceConfig.workspaceWriteKeysMapLock.Unlock()
+	wc.workspaceWriteKeysMapLock.Lock()
+	wc.writeKeyToWorkspaceIDMap = writeKeyToWorkspaceIDMap
+	wc.sourceToWorkspaceIDMap = sourceToWorkspaceIDMap
+	wc.workspaceIDToLibrariesMap = workspaceIDToLibrariesMap
+	wc.workspaceWriteKeysMapLock.Unlock()
 
 	return sourcesJSON, nil
 }
 
-func (workspaceConfig *MultiTenantWorkspacesConfig) makeHTTPRequest(
+func (wc *multiTenantWorkspacesConfig) makeHTTPRequest(
 	ctx context.Context, url string,
 ) ([]byte, int, error) {
-	req, err := Http.NewRequestWithContext(ctx, "GET", url, http.NoBody)
+	req, err := http.NewRequestWithContext(ctx, "GET", url, http.NoBody)
 	if err != nil {
 		return []byte{}, 400, err
 	}
 	// TODO: hacky way to get the backend config for multi tenant through older hosted backend config
 	if config.GetBool("BackendConfig.useHostedBackendConfig", false) {
 		req.SetBasicAuth(config.GetEnv("HOSTED_SERVICE_SECRET", ""), "")
 	} else {
-		req.SetBasicAuth(workspaceConfig.Token, "")
+		req.SetBasicAuth(wc.Token, "")
 	}
 
 	req.Header.Set("Content-Type", "application/json")
 
 	client := &http.Client{Timeout: config.GetDuration("HttpClient.timeout", 30, time.Second)}
 	resp, err := client.Do(req)
 	if err != nil {
-		return []byte{}, 400, err
+		return nil, http.StatusBadRequest, err
+	}
+
+	respBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, http.StatusBadRequest, err
 	}
 
-	var respBody []byte
-	if resp != nil && resp.Body != nil {
-		respBody, _ = IoUtil.ReadAll(resp.Body)
-		defer func() { _ = resp.Body.Close() }()
+	defer func() { _ = resp.Body.Close() }()
+
+	if resp.StatusCode >= 300 {
+		return nil, resp.StatusCode, getNotOKError(respBody, resp.StatusCode)
 	}
 
 	return respBody, resp.StatusCode, nil
```

Comment:
Question: previously we were not handling HTTP response codes properly, but we were still trying to unmarshall the response to the proper Config struct. Didn't this step fail?

Reply:
If we don't fail by checking the status code first here, we could get some unexpected results when unmarshalling as it's not reliable in some cases. For example when unmarshalling after a bad credentials response we get this: `{"message":"Invalid workspace token"}` (which is a valid JSON). That could trick the marshaller and accept it as valid.

```go
func TestAris(t *testing.T) {
	data := []byte(`{"message":"Invalid workspace token"}`)
	var sourcesJSON backendconfig.ConfigT
	err := json.Unmarshal(data, &sourcesJSON)
	require.NoError(t, err)
}
```

The above test passes.